File: AdventOfCode/1/FrequencyDevice.cpp
#include <iostream>
#include <cstdlib>
#include <unordered_set>
/**
 * Solution to https://adventofcode.com/2018/day/1
 * Author: Cody Lewis
 * Date 2018-12-01
 */

/**
 * Take the argument input and give back a table of frequency changes.
 */
int* memoize_changes(int argc, char* argv[]) {
  int* changes = (int*) malloc(sizeof(int) * (argc - 1));
  for(int i = 1; i < argc; ++i) {
    changes[i - 1] = atoi(argv[i]);
  }
  return changes;
}

/**
 * Sum up the table of changes.
 */
int find_final_frequency(int num_changes, int* changes) {
  int final_frequency = 0;
  for(int i = 0; i < num_changes; ++i) {
    final_frequency += changes[i];
  }
  return final_frequency;
}

/**
 * Continually cycle through the changes until a sum occurs twice,
 * return that sum
 */
long find_first_double(int num_changes, int* changes) {
  long current_frequency = 0;
  std::unordered_set<int>* occured_frequencies = new std::unordered_set<int>();
  while(1) {
    for(int i = 0; i < num_changes - 1; ++i) {
      current_frequency += changes[i];
      if(occured_frequencies->find(current_frequency) != occured_frequencies->end()) {
        delete(occured_frequencies);
        return current_frequency;
      }
      occured_frequencies->insert(current_frequency);
    }
  }
}

int main(int argc, char* argv[]) {
  int* changes = memoize_changes(argc, argv);
  std::cout << "The final frequency is " << find_final_frequency(argc, changes) << std::endl;
  std::cout << "The first frequency that occurs twice is " << find_first_double(argc, changes) << std::endl;
  free(changes);
  return 0;
}

File: AdventOfCode/1/FrequencyDevice.py
#!/usr/bin/env python3
'''
Solution to https://adventofcode.com/2018/day/1
Author: Cody Lewis
Date: 2018-12-01
'''


def memoize_changes():
    '''
    Create a table of the frequency changes.
    '''
    changes = []
    with open("Changes.txt") as f:
        for line in f:
            for element in line.split(" "):
                changes.append(int(element))
    return changes


def find_final_frequency(changes=memoize_changes()):
    '''
    Sum the changes together.
    '''
    final_frequency = 0
    for change in changes:
        final_frequency += change
    return final_frequency


def find_first_double(changes=memoize_changes()):
    '''
    Find a double of the sum cycle.
    '''
    running_frequency = 0
    occured_frequencies = {running_frequency : True}
    while True:
        for change in changes:
            running_frequency += change
            print(running_frequency)
            if occured_frequencies.get(running_frequency) != None:
                return running_frequency
            occured_frequencies.update({running_frequency : True})


if __name__ == "__main__":
    CHANGES = memoize_changes()
    print(f"The final frequency is {find_final_frequency(CHANGES)}")
    print(f"The first doubled frequency is {find_first_double(CHANGES)}")
